File: 61-rotate-list/rotate-list.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @param {number} k
 * @return {ListNode}
 */
var rotateRight = function(head, k) {

    if(head == null || head.next == null || k == 0){
        return head;
    }

    let length = 1;
    let cur = head;
    while(cur.next != null){
        cur = cur.next;
        length++;
    }


    cur.next = head;

    k= k % length;

    let steptoFindTail = length - k;
    let newHead = null;

    for(let i = 0; i<steptoFindTail; i++){
        cur = cur.next;
    }

    newHead = cur.next;
    cur.next = null;

    return newHead;
    
};

File: 61-rotate-list/rotate-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode rotateRight(ListNode head, int k) {
        
        if(head == null || head.next == null || k == 0){
            return head;
        }

        int len = 1;
        ListNode cur = head;
        while(cur.next != null){
            len++;
            cur = cur.next;
        }

        cur.next = head;

        k = k % len;

        int stepToFindTail = len - k;

        ListNode newhead = null;

        for(int i = 0; i<stepToFindTail; i++){
            cur = cur.next;
        }

        newhead = cur.next;

        cur.next = null;

        return newhead;
    }
}